src/Components/InMemoryEventStream.cs

```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Components
+{
+    internal sealed class InMemoryEventStream : EventStream, EventDispatcher
+    {
+        private readonly HashSet<object> _subscriptions = new HashSet<object>();
+
+        public void Dispatch<T>(T entity)
+        {
+            if (entity == null)
+                throw new ArgumentNullException(nameof(entity));
+
+            lock(_subscriptions)
```

Comment:
Was thinking the same, wouldn't the concurrent be a better option than doing the manual lock etc?